### PR TITLE
Add edge weight sample for cpu

### PR DESCRIPTION
**Pull request by husimplicity**
Add edge weight sample for cpu (homo & hetero)

### CODE REVIEW

**Comment by husimplicity** [comment]
cpu single node case to #68

**Comment by husimplicity** [comment]
distributed implementation added

**Comment by husimplicity** [comment]
Added a distributed weighted sampling link prediction example.

File: graphlearn_torch/csrc/cpu/random_sampler.cc [hunk 1]
```diff
@@ -71,6 +74,62 @@ CPURandomSampler::SampleWithEdge(const torch::Tensor& nodes, int32_t req_num) {
   return std::make_tuple(nbrs, nbrs_num, out_eid);
 }
 
+std::tuple<torch::Tensor, torch::Tensor>
+CPURandomSampler::WeightedSample(const torch::Tensor& nodes, int32_t req_num) {
```
Comment (baoleai) [inline graphlearn_torch/csrc/cpu/random_sampler.cc:78]: It's better to move this weightedsample in a separate file.

File: graphlearn_torch/csrc/cpu/random_sampler.cc [hunk 2]
```diff
@@ -175,4 +281,51 @@ void CPURandomSampler::UniformSample(const int64_t* col_begin,
   }
 }
 
+void CPURandomSampler::WeightedRandomSample(const int64_t* col_begin,
```
Comment (baoleai) [inline graphlearn_torch/csrc/cpu/random_sampler.cc:284]: WeightedRandom->Weighted cause weighted is not random.

File: graphlearn_torch/csrc/cpu/random_sampler.cc [hunk 3]
```diff
@@ -175,4 +281,51 @@ void CPURandomSampler::UniformSample(const int64_t* col_begin,
...
+                                            const int32_t req_num,
+                                            const float* prob_begin,
+                                            const float* prob_end,
+                                            int64_t* out_nbrs) {
+  // with replacement
+  const auto cap = col_end - col_begin;
+  if (req_num < cap) {
+    thread_local static std::random_device rd;
+    thread_local static std::mt19937 engine(rd());
+    std::discrete_distribution<> dist(prob_begin, prob_end);
```
Comment (baoleai) [inline graphlearn_torch/csrc/cpu/random_sampler.cc:295]: This function only works with integers. Here prob is float, It is best to use something like the alias method or reservoir algorithm for weighted sampling.

Comment (husimplicity) [reply graphlearn_torch/csrc/cpu/random_sampler.cc:295]: As shown in the test_neighbor_sampler.py, this function works well with float. The weights are set 0.1 and 0.9 for the neighbors and the sampling frequency obeys the distribution we assumed

File: graphlearn_torch/csrc/cpu/random_sampler.cc (hunk 2 above)
Comment (husimplicity) [reply graphlearn_torch/csrc/cpu/random_sampler.cc:284]: The source of randomness is from sample. The previous cases called random are actually of uniform distribution, while in this case the sampling should follow the multinomial distribution generated by the weights. Just the matter of replacing probability measure from uniform to another

File: graphlearn_torch/csrc/cpu/random_sampler.cc (hunk 1 above)
Comment (husimplicity) [reply graphlearn_torch/csrc/cpu/random_sampler.cc]: Done!

**Review by baoleai — APPROVED** [review]

**Comment by husimplicity** [comment]
solve #68

**Pull request merged**